Scrolling Waveforms blog post (#314)
mixxxdj/website: content/news/2024-01-23-improved-waveforms.md

title: Improved Scrolling Waveforms in Mixxx 2.4.0
authors: Maarten de Boer
tags: 2.4, improvements
status: draft
comments: yes
date: 2023-08-15 13:09:47

> One of the improvements of Mixxx 2.4 is a rewrite of the scrolling waveforms code. This results in smoother animation at a higher frame rate (60 fps), with less frame-drops and lower CPU load. These improvements are particularly noticeable on macOS, but also apply to Windows and Linux.

Reviewer: ```suggestion
One of the major improvements of Mixxx 2.4 is a rewrite of the scrolling waveforms code. This results in smoother animation at a higher frame rate (60 fps), with less frame-drops and lower CPU load. These improvements are particularly noticeable on macOS, but also apply to Windows and Linux.
```

The Mixxx scrolling waveforms are drawn with a number of overlapping layers: The beat grid, the looping-, intro- and outro-ranges, the markers, the end-of-track indication, and the actual waveform. Originally these layers were rendered with a combination of Qt (QPainter) and Legacy OpenGL calls, on the now deprecated QGLWidget. Profiling showed that this combination was a performance bottleneck, particularly on macOS. With the rewrite, all these layers are now using Modern OpenGL, with hardware accelerated GLSL shaders. The different waveform types (Simple, Filtered, HSV, RGB and RGB L/R) all come with a GLSL implementation. In addition to the waveforms, the spinny widgets and VU-meters have been rewritten with the same approach.

> Visually the new waveforms follow the design of the legacy waveforms, with some minor tweaks (note for example the semi-transparently filled triangles pre- and post track). What has been improved is the layout of the markers layers: Overlapping markers (multiple markers at the same beat grid location) which would previously obfuscate each other, are now automatically placed at an increasing / decreasing vertical position.

Reviewer: ```suggestion
Visually the new waveforms follow the design of the legacy waveforms, with some minor tweaks, note for example the semi-transparently filled triangles pre- and post track. What has been improved is the layout of the markers layers: Overlapping markers (multiple markers at the same beat grid location) which would previously obfuscate each other, are now automatically placed at an increasing / decreasing vertical position.
```
Because every time a bracket is opened, I try to memorize the part before it so that I can resume the previous flow after the bracket has been closed.
In contrast, a comma _mostly_ indicates a new section/level/context.

> Visually the new waveforms follow the design of the legacy waveforms, with some minor tweaks, note for example the semi-transparently filled triangles pre- and post track. What has been improved is the layout of the markers layers: Overlapping markers (multiple markers at the same beat grid location) which would previously obfuscate each other, are now automatically placed at an increasing / decreasing vertical position.

Reviewer 2: The writting is really good, but I would recommend to insert two detail screenshots showing:

- the pre-track triangles
- the stacked markers

Author: will do!

Reviewer 2: The beatgrid in the triangle screenshot has an offset. Could you adjust this?

Author: I replaced it with a different screenshot. Is this better?

The newly implemented waveforms have been beta-tested for quite a while now and are considered stable and recommended. When upgrading from older versions of Mixxx, the GLSL waveform type that best matches the old selection will be selected automatically, as well as the 60 fps frame rate. The legacy waveform types have been maintained as selectable options just in case, in the Waveforms section of the settings dialog.

> The deprecated QGLWidget has been replaced by a custom solution using a QOpenGLWindow inside a QWidget using the createWindowContainer() call. Qt also offers QOpenGLWidget to replace QGLWidget, but the QOpenGLWindow solution resulted in better performance and integrated better with the existing source code. This change will also facilitate the migration to Qt 6, planned for Mixxx 2.5.

Reviewer: ```suggestion
The deprecated QGLWidget has been replaced by a custom solution using a QOpenGLWindow inside a QWidget using the createWindowContainer() call. Qt also offers QOpenGLWidget to replace QGLWidget, but the QOpenGLWindow solution resulted in better performance and integrated better with the existing source code. This change will also facilitate the migration to Qt 6, planned for Mixxx 2.5.
```
Haha, looks exactly the same in the diff..
This is some &nbsp; char from https://stackoverflow.com/a/54174466, just so 2.5 doesn't hang around alone on the last line.

Reviewer: such stuff should probably better be dealt with by some config

> Finally, a last minute addition is an alternative mode to synchronise the scrolling waveform animation with the display refresh rate, using a so-called phase-locked-loop (PLL), which attempts to track the actual refresh rate and timing automatically. On particular hardware, the default periodic timer-based approach can result in jitter and frame drops and the PLL may gives better results. See [instructions to activate this mode](https://github.com/mixxxdj/mixxx/wiki/Activating-Phase%E2%80%90Locked%E2%80%90Loop-VSync-Mode-for-Scrolling-Waveforms) to activate this mode.

Reviewer: ```suggestion
Finally, a last minute addition is an alternative mode to synchronise the scrolling waveform animation with the display refresh rate, using a so-called phase-locked-loop (PLL), which attempts to track the actual refresh rate and timing automatically. On particular hardware, the default periodic timer-based approach can result in jitter and frame drops and the PLL may gives better results. See [instructions to activate this mode](https://github.com/mixxxdj/mixxx/wiki/Activating-Phase%E2%80%90Locked%E2%80%90Loop-VSync-Mode-for-Scrolling-Waveforms).
```

title: Improved Scrolling Waveforms in Mixxx 2.4.0
authors: Maarten de Boer
tags: 2.4, improvements
status: draft
comments: yes
date: 2023-08-15 13:09:47

One of the major improvements of Mixxx 2.4 is a revision of the scrolling waveform implementation, to achieve smoother animation at a higher frame rate (60 fps), with less frame-drops and lower CPU load.

> ## Rewrite

Reviewer 3: Not sure, do we really need this headline? It looks bad in the news section of the homepage.

I think we can either drop this headline, make the paragraph before the headline longer so that it gets pushed out of the excerpt or specify an excerpt explicitly in the metadata.

![RGB L/R (GLSL) Waveform type with pre-track trianges]({static}/images/news/glsl-rgb-lr-waveform.png)

The newly implemented GLSL waveforms have been beta-tested for several months now and are considered stable and recommended. The old waveform types remain available, just in case, as an option in the Waveforms section of the settings dialog, marked with "(legacy)", and might be removed in future versions. If you experience issues with the new waveform types, you are encouraged to file a bug report!

## Improved marker layout

Additionally, the markers layer has been improved: Multiple markers at the same beat grid location or sample position that would previously overlap and obscure each other, are now automatically stacked, placed at a respectively increased or decreased vertical offset. This applies to (hot) cue markers, intro- and outro-markers and loop-markers, and should be a big improvement for DJs that make heavy use of these markers.

![Overlapping markers drawn at different vertical positions]({static}/images/news/overlapping-markers.png)

## Some technical insights

> Mixxx uses the Qt software framework for its user interface. Originally the layers that form the scrolling waveform display were rendered on the now deprecated QGLWidget through a combination of 'QPainter' and Legacy OpenGL function calls. Profiling showed that this combination was a performance bottleneck, particularly on macOS.

Reviewer 3: ```suggestion
Mixxx uses the Qt software framework for its user interface. Originally the layers that form the scrolling waveform display were rendered on the now deprecated `QGLWidget` through a combination of `QPainter` and legacy OpenGL function calls. Profiling showed that this combination was a performance bottleneck, particularly on macOS.
```

With the rewrite, all the layers mentioned above are now implemented using Modern OpenGL code: All different waveform types, i.e, Simple, Filtered, HSV, RGB and RGB L/R, are now hardware accelarated using GLSL shaders. In addition to the waveforms, the spinny widgets and VU-meters code has been revised with the same approach.

> The deprecated QGLWidget has been replaced by a custom solution using a QOpenGLWindow inside a QWidget using the createWindowContainer() call. Qt also offers QOpenGLWidget to replace QGLWidget, but the QOpenGLWindow solution resulted in better performance and integrated better with the existing source code. This change will also facilitate the migration to Qt 6, planned for Mixxx 2.5.

Reviewer 3: ```suggestion
The deprecated `QGLWidget` has been replaced by a custom solution using a `QOpenGLWindow` inside a `QWidget` using the `createWindowContainer()` call. Qt also offers `QOpenGLWidget` to replace `QGLWidget`, but the `QOpenGLWindow` solution resulted in better performance and integrated better with the existing source code. This change will also facilitate the migration to Qt 6, planned for Mixxx 2.5.
```

## New display synchronization mode

> Finally, a last minute addition is an alternative mode to synchronise the scrolling waveform animation with the display refresh rate, using a so-called phase-locked-loop (PLL). This mechanism attempts to track the actual refresh rate and timing of the display automatically. On particular hardware, the default periodic timer-based approach can result in jitter and frame drops, and the PLL may give better results. The PLL has been made the default on macOS. See [instructions to change the V-Sync mode](https://github.com/mixxxdj/mixxx/wiki/Changing-the-VSync-Mode-for-Scrolling-Waveforms) if you want to try which mode works best for you.

Reviewer 4: ```suggestion
To avoid micro jitter, it is required to display the rendered picture in the right moment. This is achieved by synchronising the scrolling waveform animation with the display refresh rate.  As a last minute addition an alternative mode to using a so-called phase-locked-loop (PLL) has been introduced. This mechanism attempts to track the actual refresh rate and timing of the display automatically. On particular hardware, the default periodic timer-based approach can result in jitter and frame drops, and the PLL may give better results. The PLL has been made the default on macOS other targets did not show an improvement. See [instructions to change the V-Sync mode](https://github.com/mixxxdj/mixxx/wiki/Changing-the-VSync-Mode-for-Scrolling-Waveforms) if you want to try which mode works best for you. This setting does not effect the audio control performance. 
```
